src/deluge/gui/ui/keyboard/layout/column_controls.cpp

```diff
@@ -0,0 +1,461 @@
+};
+
+const char* functionNames[] = {
+    /* VELOCITY    */ "VEL",
+    /* MOD         */ "MOD",
+    /* CHORD       */ "CHRD",
+    /* CHORD_MEM   */ "CMEM",
+    /* SCALE_MODE  */ "SMOD",
+    /* BEAT_REPEAT */ "BEAT",
+};
+
+void ColumnControlsKeyboard::enableNote(uint8_t note, uint8_t velocity) {
+	currentNotesState.enableNote(note, velocity);
+	for (int i = 0; i < MAX_CHORD_NOTES; i++) {
+		auto offset = chordSemitoneOffsets[i];
+		if (!offset) {
+			break;
+		}
+		currentNotesState.enableNote(note + offset, velocity, true);
+	}
+}
+
+void ColumnControlsKeyboard::evaluatePads(PressedPad presses[kMaxNumKeyboardPadPresses]) {
+	char modelStackMemory[MODEL_STACK_MAX_SIZE];
+	ModelStack* modelStack = setupModelStackWithSong(modelStackMemory, currentSong);
+	ModelStackWithTimelineCounter* modelStackWithTimelineCounter =
+	    modelStack->addTimelineCounter(currentSong->currentClip);
+
+	leftColMinHeld = false;
+	leftColMaxHeld = false;
+
+	rightColMinHeld = false;
+	rightColMaxHeld = false;
+
+	for (int32_t idxPress = 0; idxPress < kMaxNumKeyboardPadPresses; ++idxPress) {
+		auto pressed = presses[idxPress];
+		// in note columns
+		if (pressed.x == LEFT_COL) {
+			if (pressed.active) {
+				if (pressed.y == 0) {
+					leftColMinHeld = true;
+				}
+				else if (pressed.y == 7) {
+					leftColMaxHeld = true;
+				}
+			}
+			else if (horizontalScrollingLeftCol && pressed.y == 7) {
+				handlePad(modelStackWithTimelineCounter, leftColPrev, pressed);
+				horizontalScrollingLeftCol = false;
+				continue;
+			}
+			if (!horizontalScrollingLeftCol) {
+				handlePad(modelStackWithTimelineCounter, leftCol, pressed);
+			}
+		}
+		else if (pressed.x == RIGHT_COL) {
+			if (pressed.active) {
+				if (pressed.y == 0) {
+					rightColMinHeld = true;
+				}
+				else if (pressed.y == 7) {
+					rightColMaxHeld = true;
+				}
+			}
+			else if (horizontalScrollingRightCol && pressed.y == 7) {
+				handlePad(modelStackWithTimelineCounter, rightColPrev, pressed);
+				horizontalScrollingRightCol = false;
+				continue;
+			}
+			if (!horizontalScrollingRightCol) {
+				handlePad(modelStackWithTimelineCounter, rightCol, pressed);
+			}
+		}
+	}
+}
+
+void ColumnControlsKeyboard::handlePad(ModelStackWithTimelineCounter* modelStackWithTimelineCounter,
+                                       ColumnControlFunction func, PressedPad pad) {
+	switch (func) {
+	case VELOCITY:
+		if (pad.active) {
+			vDisplay = velocityMin + pad.y * velocityStep;
+			velocity = (vDisplay + kHalfStep) >> kVelModShift;
+			display->displayPopup(velocity);
+		}
+		else if (!pad.padPressHeld) {
+			velocity32 = velocityMin + pad.y * velocityStep;
+			velocity = (velocity32 + kHalfStep) >> kVelModShift;
+		}
+		else {
+			vDisplay = velocity32;
+			velocity = (velocity32 + kHalfStep) >> kVelModShift;
+		}
+		break;
+	case MOD:
+		if (pad.active) {
+			modDisplay = modMin + pad.y * modStep;
+			getCurrentInstrument()->processParamFromInputMIDIChannel(CC_NUMBER_Y_AXIS, modDisplay,
```

Comment:
should be CC1 (mod wheel), this records it as MPE y to the currently sounding notes instead

Reply:
Actually looking again I think this is right as is, incoming CC1 is converted to this value. This should be fine